# Load in data 
+ Data comes as a tar file
+ I opted to just load in files into pandas instead of attempting to unwrap that tar file in python
+ We have a file containing all of the edges facebook.csv
+ What we need to do is combine the .feat files with the .featnames files
+ below you can see what each of these files looks like

In [1]:
import pandas as pd 

In [2]:



example= pd.read_csv('0.feat',sep=" ",header=None)
print(example.shape)
example.head()


(347, 225)


0    1    2    3    4    5    6    7    8    9   ...   215  216  217  218  \
0    1    0    0    0    0    0    0    0    0    0 ...     0    0    0    0   
1    2    0    0    0    0    0    0    0    0    0 ...     0    0    0    0   
2    3    0    0    0    0    0    0    0    1    0 ...     0    0    0    1   
3    4    0    0    0    0    0    0    0    0    0 ...     0    0    0    0   
4    5    0    0    0    0    0    0    0    0    0 ...     0    0    0    0   

   219  220  221  222  223  224  
0    0    0    0    0    0    0  
1    0    0    0    0    0    0  
2    0    0    0    0    0    0  
3    0    0    0    0    0    0  
4    0    0    0    0    0    0  

[5 rows x 225 columns]

In [3]:

example= pd.read_csv('0.featnames',sep=" ",header=None)
print(example.shape)
example.head()

(224, 4)


0                    1        2  3
0  0  birthday;anonymized  feature  0
1  1  birthday;anonymized  feature  1
2  2  birthday;anonymized  feature  2
3  3  birthday;anonymized  feature  3
4  4  birthday;anonymized  feature  4

# structure of our data

+  Column values in the .feat files(225 columns) are the equivalent index value for our .featnames files(224 rows)
    + The difference of 1 in the shape is due to the 1st column in the .feat file representing our node value and not a feature value
+ First goal is to join the names from .featnames in the feat files


### Build names lists
+ our files are stored as a random int.name
    + I.E  '0.featnames' and '0.feat'
     + Below we use list comprehension to build out file names


In [4]:
names=["0","107","1684","1912","3437",'348',"3980","414","686","698"]
feat_file_names= [x+".feat" for x in names]
featnames_filenames=[x+".featnames" for x in names]

In [5]:
# check names
featnames_filenames

['0.featnames',
 '107.featnames',
 '1684.featnames',
 '1912.featnames',
 '3437.featnames',
 '348.featnames',
 '3980.featnames',
 '414.featnames',
 '686.featnames',
 '698.featnames']

### Load files into pandas 

In [6]:
list_feat_names=[]
list_features=[]
for i,x in enumerate(featnames_filenames):
    list_feat_names.append(pd.read_csv(x,sep=" ",header=None))
    list_features.append(pd.read_csv(feat_file_names[i],sep=" ",header=None))

### Check shape matches earlier shape 

In [7]:
list_feat_names[0].shape

(224, 4)

In [8]:
list_features[0].shape

(347, 225)

### Practice on one set of each document

In [224]:
df=list_features[0]
df.head()
#df = df.drop(df.columns[[0, 1]], axis=1)
indexes=df.iloc[:,0]
df=df.iloc[:,1:]
df.columns= list_feat_names[0][1]
df["node"]=indexes
df.head()


1  birthday;anonymized  birthday;anonymized  birthday;anonymized  \
0                    0                    0                    0   
1                    0                    0                    0   
2                    0                    0                    0   
3                    0                    0                    0   
4                    0                    0                    0   

1  birthday;anonymized  birthday;anonymized  birthday;anonymized  \
0                    0                    0                    0   
1                    0                    0                    0   
2                    0                    0                    0   
3                    0                    0                    0   
4                    0                    0                    0   

1  birthday;anonymized  birthday;anonymized  education;classes;id;anonymized  \
0                    0                    0                                0   
1                    0                    0                                0   
2                    0                    1                                0   
3                    0                    0                                0   
4                    0                    0                                0   

1  education;classes;id;anonymized  ...   work;start_date;anonymized  \
0                                0  ...                            0   
1                                0  ...                            0   
2                                0  ...                            0   
3                                0  ...                            0   
4                                0  ...                            0   

1  work;start_date;anonymized  work;start_date;anonymized  \
0                           0                           0   
1                           0                           0   
2                           0                           1   
3                           0                           0   
4                           0                           0   

1  work;start_date;anonymized  work;start_date;anonymized  \
0                           0                           0   
1                           0                           0   
2                           0                           0   
3                           0                           0   
4                           0                           0   

1  work;start_date;anonymized  work;start_date;anonymized  \
0                           0                           0   
1                           0                           0   
2                           0                           0   
3                           0                           0   
4                           0                           0   

1  work;start_date;anonymized  work;with;id;anonymized  node  
0                           0                        0     1  
1                           0                        0     2  
2                           0                        0     3  
3                           0                        0     4  
4                           0                        0     5  

[5 rows x 225 columns]

In [78]:
# final_product=[]
# for i,x in enumerate(list_features):
#     df=x
#     indexes=x.iloc[:,0]
#     df=x.iloc[:,1:]
#     df.columns= list_feat_names[i][1]
#     df["nodes"]=indexes
#     final_product.append(df)



final_product2=[]
for i,x in enumerate(list_features):
    df=x
    indexes=x.iloc[:,0]
    df=x.iloc[:,1:]
    df.columns= list_feat_names[i][3].astype(int)
    df.index=indexes
    final_product2.append(df)
    
    


In [9]:
final_product2=[]
for i,x in enumerate(list_features):
    df=x
    indexes=x.iloc[:,0]
    df=x.iloc[:,1:]
    df.columns= list_feat_names[i][3].astype(int)
    df.index=indexes
    final_product2.append(df)

In [130]:
index_list=[]
duplicates_index=
for x in final_product2:

    for index in x.index:
        if index==1758:
            index_list.append(index)
    print (index_list)
            

[]
[]
[]
[1913]
[]
[]
[]
[]
[]
[]


In [295]:
index_list=[]
duplicates_index=[]
total_rows=0
for x in final_product:
    for index in x.index:
        total_rows+=1
        if index not in index_list:
            index_list.append(index)
        else:
            duplicates_index.append(index)


len(duplicates_index)

132

In [141]:
print(total_rows)
len(duplicates_index)

4167-132

## should be 4038 + 3 egos 

4167


4035

In [157]:


# total_rows=0
# df_count=0
# for x in final_product2:
#     print("values are in df "+str(df_count))
#     index_list=[]
#     duplicates_index=[]
#     for col in x.columns:
#         if col not in index_list:
#             index_list.append(col)
#         else:
#             duplicates_index.append(col)
#             print(col)
#     df_count+=1
# final_product2[0].columns.str

# [col for col in final_product2[0].columns if col==88]

values are in df 0
88
50
52
129
132
137
138
157
158
160
162
164
165
166
168
169
170
171
values are in df 1
110
84
298
264
554
228
451
230
485
518
247
248
536
52
538
128
129
609
561
281
133
84
176
134
178
615
283
617
618
619
264
157
672
674
675
678
196
679
161
162
164
199
166
200
201
168
169
170
171
681
683
values are in df 2
935
176
1095
617
284
322
538
1144
935
176
617
157
294
158
197
164
166
201
168
170
values are in df 3
894
558
902
134
907
908
442
852
863
52
558
128
612
933
82
281
84
176
134
908
619
157
310
976
160
161
162
723
198
164
199
166
726
680
168
202
169
170
values are in df 4
297
1223
619
52
537
1210
128
129
296
174
176
1220
179
283
619
138
157
723
198
199
values are in df 5
128
84
225
52
128
84
157
values are in df 6
52
157
values are in df 7
128
84
52
84
157
values are in df 8
359
343
359
157
values are in df 9
359
366
366
157


[88, 88]

In [176]:
final_product=[]
for i,x in enumerate(list_features):
    df=x
    indexes=x.iloc[:,0]
    df=x.iloc[:,1:]
    df.columns= list_feat_names[i][1].astype(str)+'_'+list_feat_names[i][3].astype(str)
  #  df["nodes"]=indexes
    df.index=indexes
    final_product.append(df)
    
    

In [203]:
my_nodes={}
for node in range(1,4036):
    my_nodes.setdefault(node, [])



for x in final_product:
    n=0
    for node in range(1,4036):
        for index in x.index:
            if index==node:
                my_nodes[node].append(x.loc[node,])
                



In [181]:
index_list=[]
duplicates_index=[]
total_rows=0
for x in final_product:
    for index in x.index:
        total_rows+=1
        if index not in index_list:
            index_list.append(index)
        else:
            duplicates_index.append(index)


len(duplicates_index)

132

In [188]:
col_list=[]
duplicates_index=[]
total_rows=0
for x in final_product:
    for col in x.columns:
        if col not in col_list:
            col_list.append(col)
        else:
            duplicates_index.append(col)


In [193]:
df = pd.DataFrame(np.nan, index=np.arange(1,4040), columns=col_list)

In [ ]:

new_df=pd.DataFrame(data = [my_nodes[1][0].values] * len(my_nodes[1][0]), columns = my_nodes[1][0].index)
df.merge(new_df, left_index=True, right_index=True)
         

In [251]:
my_nodes[1][0].index
my_nodes[1][0].values

Index(['birthday;anonymized_0', 'birthday;anonymized_1',
       'birthday;anonymized_2', 'birthday;anonymized_3',
       'birthday;anonymized_4', 'birthday;anonymized_5',
       'birthday;anonymized_6', 'birthday;anonymized_7',
       'education;classes;id;anonymized_8',
       'education;classes;id;anonymized_9',
       ...
       'work;start_date;anonymized_200', 'work;start_date;anonymized_201',
       'work;start_date;anonymized_168', 'work;start_date;anonymized_202',
       'work;start_date;anonymized_169', 'work;start_date;anonymized_170',
       'work;start_date;anonymized_171', 'work;start_date;anonymized_203',
       'work;start_date;anonymized_204', 'work;with;id;anonymized_205'],
      dtype='object', length=224)

# Prove col names are same

In [267]:
my_list=[x for x in my_nodes[1][0].index]
my_list2= [x for x in df.iloc[0,:].index]

for x in my_list:
    if x in my_list2:
        print(x)

# Build blank df of NAN


In [279]:

df = pd.DataFrame(np.nan, index=np.arange(1,4040), columns=col_list)

# still creating appendixed cols

In [278]:
df.merge(my_nodes[1][0].to_frame().T, left_index=True, right_index=True)

birthday;anonymized_0_x  birthday;anonymized_1_x  birthday;anonymized_2_x  \
1                      NaN                      NaN                      NaN   

   birthday;anonymized_3_x  birthday;anonymized_4_x  birthday;anonymized_5_x  \
1                      NaN                      NaN                      NaN   

   birthday;anonymized_6_x  birthday;anonymized_7_x  \
1                      NaN                      NaN   

   education;classes;id;anonymized_8_x  education;classes;id;anonymized_9_x  \
1                                  NaN                                  NaN   

               ...                work;start_date;anonymized_200_y  \
1              ...                                               0   

   work;start_date;anonymized_201_y  work;start_date;anonymized_168_y  \
1                                 0                                 0   

   work;start_date;anonymized_202_y  work;start_date;anonymized_169_y  \
1                                 0                                 0   

   work;start_date;anonymized_170_y  work;start_date;anonymized_171_y  \
1                                 0                                 0   

   work;start_date;anonymized_203_y  work;start_date;anonymized_204_y  \
1                                 0                                 0   

   work;with;id;anonymized_205_y  
1                              0  

[1 rows x 1630 columns]

In [275]:
my_nodes[1][0].to_frame().T

birthday;anonymized_0  birthday;anonymized_1  birthday;anonymized_2  \
1                      0                      0                      0   

   birthday;anonymized_3  birthday;anonymized_4  birthday;anonymized_5  \
1                      0                      0                      0   

   birthday;anonymized_6  birthday;anonymized_7  \
1                      0                      0   

   education;classes;id;anonymized_8  education;classes;id;anonymized_9  \
1                                  0                                  0   

              ...               work;start_date;anonymized_200  \
1             ...                                            0   

   work;start_date;anonymized_201  work;start_date;anonymized_168  \
1                               0                               0   

   work;start_date;anonymized_202  work;start_date;anonymized_169  \
1                               0                               0   

   work;start_date;anonymized_170  work;start_date;anonymized_171  \
1                               0                               0   

   work;start_date;anonymized_203  work;start_date;anonymized_204  \
1                               0                               0   

   work;with;id;anonymized_205  
1                            0  

[1 rows x 224 columns]

In [284]:
new_df= pd.concat([final_product[0],final_product[1],final_product[2],final_product[3],final_product[4],final_product[5],final_product[6],final_product[7],final_product[8],final_product[9]])

E:\Anaconda\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """Entry point for launching an IPython kernel.


In [290]:
df_9=pd.concat([final_product[0],final_product[1],final_product[2]])

E:\Anaconda\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """Entry point for launching an IPython kernel.


In [291]:
df_9

birthday;anonymized_0  birthday;anonymized_1  birthday;anonymized_1003  \
0                                                                              
1                       0.0                    0.0                       NaN   
2                       0.0                    0.0                       NaN   
3                       0.0                    0.0                       NaN   
4                       0.0                    0.0                       NaN   
5                       0.0                    0.0                       NaN   
6                       0.0                    1.0                       NaN   
7                       0.0                    0.0                       NaN   
8                       0.0                    0.0                       NaN   
9                       0.0                    0.0                       NaN   
10                      0.0                    0.0                       NaN   
11                      0.0                    0.0                       NaN   
12                      0.0                    0.0                       NaN   
13                      0.0                    0.0                       NaN   
14                      0.0                    0.0                       NaN   
15                      0.0                    0.0                       NaN   
16                      0.0                    0.0                       NaN   
17                      1.0                    0.0                       NaN   
18                      0.0                    0.0                       NaN   
19                      0.0                    0.0                       NaN   
20                      0.0                    1.0                       NaN   
21                      0.0                    0.0                       NaN   
22                      0.0                    0.0                       NaN   
23                      0.0                    0.0                       NaN   
24                      0.0                    0.0                       NaN   
25                      0.0                    0.0                       NaN   
26                      0.0                    0.0                       NaN   
27                      0.0                    0.0                       NaN   
28                      0.0                    0.0                       NaN   
29                      0.0                    0.0                       NaN   
30                      0.0                    0.0                       NaN   
...                     ...                    ...                       ...   
3407                    NaN                    NaN                       0.0   
3408                    NaN                    NaN                       0.0   
3409                    NaN                    NaN                       0.0   
3410                    NaN                    NaN                       0.0   
3411                    NaN                    NaN                       0.0   
3412                    NaN                    NaN                       0.0   
3413                    NaN                    NaN                       0.0   
3414                    NaN                    NaN                       0.0   
3415                    NaN                    NaN                       0.0   
3416                    NaN                    NaN                       0.0   
3417                    NaN                    NaN                       0.0   
3418                    NaN                    NaN                       0.0   
3419                    NaN                    NaN                       0.0   
3420                    NaN                    NaN                       0.0   
3421                    NaN                    NaN                       0.0   
3422                    NaN                    NaN                       0.0   
3423                    NaN                    NaN                       0.0   
3424  

In [293]:
my_nodes[88]

[birthday;anonymized_0                       0
 birthday;anonymized_1                       0
 birthday;anonymized_2                       0
 birthday;anonymized_3                       0
 birthday;anonymized_4                       0
 birthday;anonymized_5                       0
 birthday;anonymized_6                       0
 birthday;anonymized_7                       0
 education;classes;id;anonymized_8           0
 education;classes;id;anonymized_9           0
 education;classes;id;anonymized_10          0
 education;classes;id;anonymized_11          0
 education;classes;id;anonymized_12          0
 education;concentration;id;anonymized_13    0
 education;concentration;id;anonymized_14    0
 education;concentration;id;anonymized_15    0
 education;concentration;id;anonymized_16    0
 education;concentration;id;anonymized_17    0
 education;concentration;id;anonymized_18    0
 education;concentration;id;anonymized_19    0
 education;degree;id;anonymized_20           0
 education;de

In [294]:
duplicates_index

['birthday;anonymized_0',
 'birthday;anonymized_1',
 'birthday;anonymized_2',
 'birthday;anonymized_3',
 'birthday;anonymized_4',
 'birthday;anonymized_5',
 'birthday;anonymized_6',
 'birthday;anonymized_7',
 'education;concentration;id;anonymized_13',
 'education;concentration;id;anonymized_14',
 'education;concentration;id;anonymized_15',
 'education;concentration;id;anonymized_18',
 'education;degree;id;anonymized_21',
 'education;degree;id;anonymized_22',
 'education;degree;id;anonymized_23',
 'education;school;id;anonymized_44',
 'education;school;id;anonymized_52',
 'education;type;anonymized_53',
 'education;type;anonymized_54',
 'education;type;anonymized_55',
 'education;year;id;anonymized_58',
 'education;year;id;anonymized_59',
 'education;year;id;anonymized_60',
 'education;year;id;anonymized_61',
 'education;year;id;anonymized_62',
 'education;year;id;anonymized_63',
 'education;year;id;anonymized_64',
 'education;year;id;anonymized_65',
 'education;year;id;anonymized_66',

In [296]:
duplicates_index

[58,
 171,
 990,
 58,
 1419,
 107,
 1171,
 171,
 1405,
 1450,
 1656,
 1505,
 1534,
 1642,
 1666,
 1758,
 1726,
 58,
 136,
 1465,
 428,
 1577,
 1718,
 563,
 1085,
 353,
 363,
 366,
 34,
 376,
 389,
 198,
 414,
 420,
 428,
 475,
 483,
 484,
 517,
 526,
 538,
 173,
 563,
 566,
 567,
 580,
 373,
 465,
 378,
 107,
 594,
 596,
 391,
 394,
 400,
 601,
 412,
 606,
 423,
 34,
 427,
 428,
 431,
 348,
 434,
 436,
 363,
 438,
 483,
 514,
 629,
 461,
 553,
 370,
 637,
 475,
 480,
 641,
 496,
 492,
 374,
 649,
 500,
 651,
 506,
 376,
 515,
 524,
 567,
 542,
 544,
 422,
 173,
 395,
 556,
 558,
 559,
 561,
 513,
 563,
 566,
 698,
 810,
 857,
 811,
 860,
 769,
 840,
 862,
 729,
 697,
 708,
 776,
 819,
 703,
 800,
 828,
 830,
 713,
 745,
 747,
 798,
 753,
 823,
 772,
 719,
 774,
 825,
 803,
 804,
 805,
 856]

In [297]:
my_nodes[58]

[birthday;anonymized_0                       0
 birthday;anonymized_1                       0
 birthday;anonymized_2                       0
 birthday;anonymized_3                       0
 birthday;anonymized_4                       0
 birthday;anonymized_5                       0
 birthday;anonymized_6                       0
 birthday;anonymized_7                       0
 education;classes;id;anonymized_8           0
 education;classes;id;anonymized_9           0
 education;classes;id;anonymized_10          0
 education;classes;id;anonymized_11          0
 education;classes;id;anonymized_12          0
 education;concentration;id;anonymized_13    0
 education;concentration;id;anonymized_14    1
 education;concentration;id;anonymized_15    0
 education;concentration;id;anonymized_16    0
 education;concentration;id;anonymized_17    0
 education;concentration;id;anonymized_18    0
 education;concentration;id;anonymized_19    0
 education;degree;id;anonymized_20           0
 education;de

In [299]:
for x in final_product:
    print(x.shape)

(347, 224)
(1045, 576)
(792, 319)
(755, 480)
(547, 262)
(227, 161)
(59, 42)
(159, 105)
(170, 63)
(66, 48)
